### PR TITLE
feat: kakarot contract wrapper

### PR DESCRIPTION
**Pull request by greged93**
Add a `KakarotContract` struct to wrap any starknet calls to kakarot.

Time spent on this PR: 0.3 day

Resolves: #242

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [x] Testing

# What is the new behavior?
Refactor `KakarotClient` in order to use a instance of the new structure `KakarotContract` for all starknet calls to kakarot. Add tests for vec_felt_to_bytes.

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: crates/core/src/client/mod.rs [hunk 1]
```diff
@@ -83,41 +82,19 @@ impl<P: Provider + Send + Sync> KakarotEthApi<P> for KakarotClient<P> {
...
-        let contract_bytecode_in_u8: Vec<u8> = contract_bytecode.into_iter().flat_map(|x| x.to_bytes_be()).collect();
-        let bytes_result = Bytes::from(contract_bytecode_in_u8);
-        Ok(bytes_result)
+        let bytecode = self
+            .kakarot_contract
+            .bytecode(&self.starknet_provider, &starknet_contract_address, &starknet_block_id)
+            .await?;
+
+        // TODO test the bytecode is correct
+        let bytecode: Bytes = vec_felt_to_bytes(bytecode);
```
Comment (ClementWalter) [inline crates/core/src/client/mod.rs:96]: and this could be in the `AccountContract` indeed

File: crates/core/src/client/mod.rs [hunk 2]
```diff
@@ -133,31 +110,22 @@ impl<P: Provider + Send + Sync> KakarotEthApi<P> for KakarotClient<P> {
...
+            .eth_call(&self.starknet_provider, &ethereum_address, &mut calldata, &starknet_block_id)
+            .await?;
 
         // Parse and decode Kakarot's call return data (temporary solution and not scalable - will
         // fail is Kakarot API changes)
         // Declare Vec of Result
         // TODO: Change to decode based on ABI or use starknet-rs future feature to decode return
         // params
-        let return_data = decode_eth_call_return(&call_result)?;
+        let return_data = decode_eth_call_return(&result)?;
```
Comment (ClementWalter) [inline crates/core/src/client/mod.rs:125]: could we add the decoding in the `KakarotContract` itself?

File: crates/core/src/client/mod.rs [hunk 3]
```diff
@@ -395,25 +363,9 @@ impl<P: Provider + Send + Sync> KakarotEthApi<P> for KakarotClient<P> {
...
-            .ok_or_else(|| DataDecodingError::InvalidReturnArrayLength {
-                entrypoint: "balance".into(),
-                expected: 1,
-                actual: 0,
-            })?
-            .to_bytes_be();
-
-        let balance = U256::from_be_bytes(balance);
+        let balance: Felt252Wrapper =
+            self.kakarot_contract.balance(&self.starknet_provider, &starknet_address, &starknet_block_id).await?.into();
```
Comment (ClementWalter) [inline crates/core/src/client/mod.rs:367]: like for the `AccountContract`, this doesn't really belong to Kakarot, but we definitely need something like a `FeeTokenContract` or just `EthContract`.

File: crates/core/src/client/mod.rs [hunk 4]
```diff
@@ -563,30 +515,13 @@ impl<P: Provider + Send + Sync> KakarotStarknetApi<P> for KakarotClient<P> {
...
-            .to_bytes_be();
-
-        // Workaround as .get(12..32) does not dynamically size the slice
-        let slice: &[u8] = evm_address
-            .get(12..32)
-            .ok_or_else(|| ConversionError::Other("error converting from [u8; 32] to &[u8]".into()))?;
-
-        let evm_address: [u8; 20] = slice.try_into().unwrap(); // safe unwrap as slice is of size 20
+        let evm_address: Felt252Wrapper = self
+            .kakarot_contract
```
Comment (ClementWalter) [inline crates/core/src/client/mod.rs:519]: `AccountContract` then here

File: crates/core/src/kakarot/mod.rs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    _phantom: PhantomData<P>,
+}
+
+impl<P: Provider + Send + Sync> KakarotContract<P> {
+    #[must_use]
+    pub fn new(address: FieldElement, proxy_account_class_hash: FieldElement) -> Self {
+        Self { address, proxy_account_class_hash, _phantom: PhantomData }
+    }
+
+    pub async fn balance(
```
Comment (ClementWalter) [inline crates/core/src/kakarot/mod.rs:24]: wdyt about creating a `EthContract` on the same model for this `balanceOf` ?

File: crates/core/src/kakarot/mod.rs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        // Make the function call to get the balance
+        let result = provider.call(request, block_id).await?;
+        Ok(*result.first().ok_or_else(|| DataDecodingError::InvalidReturnArrayLength {
+            entrypoint: "balance".into(),
+            expected: 1,
+            actual: 0,
+        })?)
+    }
+
+    pub async fn bytecode(
```
Comment (ClementWalter) [inline crates/core/src/kakarot/mod.rs:47]: and `AccountContract` here

File: crates/core/src/kakarot/mod.rs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+use crate::client::errors::EthApiError;
+use crate::client::helpers::DataDecodingError;
+
+pub struct KakarotContract<P> {
+    pub address: FieldElement,
+    pub proxy_account_class_hash: FieldElement,
+    _phantom: PhantomData<P>,
+}
+
+impl<P: Provider + Send + Sync> KakarotContract<P> {
```
Comment (ClementWalter) [inline crates/core/src/kakarot/mod.rs:18]: let's make this `impl` returned parsed results so we don't do conversation outside of the contract.

**Review by ClementWalter — COMMENTED** [review]
I propose to create more! ie a AccountContract for calling (C/EO)Account and a EthContract for the fee token contract. Then these contracts `impl` should return parsed results so as to ease their use elsewhere in the app

File: crates/core/src/kakarot/mod.rs (hunk 5 above)
Comment (greged93) [reply crates/core/src/kakarot/mod.rs:24]: Should I do `EthContract` or `Erc20Contract`?

Comment (ClementWalter) [reply crates/core/src/kakarot/mod.rs:24]: `Erc20Contract` indeed

File: crates/core/src/client/mod.rs (hunk 2 above)
Comment (greged93) [reply crates/core/src/client/mod.rs:125]: I'm wondering if it makes sense to add this decoding everywhere. Like should it be the responsibility of the `Kakarot` contract to decode from `Vec<FieldElement>` to `Bytes`?

Comment (ClementWalter) [reply crates/core/src/client/mod.rs:125]: I think so, when you use ethers-rs for example, you expect the `contract` object to return parsed data, not the raw bytecode. In our case, I think that eth_call is supposed to return the EVM response to a eth_call, ie a `Bytes` ? (or whatever type we use in the rest of the app to manipulate such an object)

File: crates/core/src/client/mod.rs (hunk 3 above)
Comment (greged93) [reply crates/core/src/client/mod.rs]: issue #289 created for this

**Review by ClementWalter — APPROVED** [review]

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**